make/gotools.mk

```diff
@@ -0,0 +1,126 @@
+# tool.
+# In contexts where `$(VARNAME)` is not available, such as when invoking external shell scripts from the Makefile, the
+# tools can also be accessed by their basename without a path prefix, as `$(GOTOOLS_BIN)` is added to the `PATH` when
+# including this file.
+#
+# This file also defines two static, global targets:
+#   gotools-clean: this removes all gotools-related data
+#   gotools-all: this builds all gotools.
+
+GOTOOLS_PROJECT_ROOT ?= $(CURDIR)
+GOTOOLS_ROOT ?= $(GOTOOLS_PROJECT_ROOT)/.gotools
+GOTOOLS_BIN ?= $(GOTOOLS_ROOT)/bin
+
+PATH := $(GOTOOLS_BIN):$(PATH)
+export PATH
+
+_GOTOOLS_ALL_GOTOOLS :=
+
+define go-tool-impl
+# The variable via which the executable can be referenced
+_gotools_var_name := $(strip $(1))
+# The importable Go package path that contains the "main" package for the tool
+_gotools_pkg := $(firstword $(subst @, ,$(strip $(2))))
+# The version of the tool (if a version was explicitly specified)
+_gotools_version := $(word 2,$(subst @, ,$(strip $(2))))
+# The folder containing the go.mod file, if not the root folder
+ifeq ($(strip $(3)),)
+_gotools_mod_root := $(GOTOOLS_PROJECT_ROOT)
+else
+_gotools_mod_root := $(strip $(3))
+endif
+
+# We need to strip a `/v2` (etc.) suffix to derive the tool binary's basename.
+_gotools_bin_name := $$(notdir $$(shell echo "$$(_gotools_pkg)" | sed -E 's@/v[[:digit:]]+$$$$@@g'))
+_gotools_canonical_bin_path := $(GOTOOLS_BIN)/$$(_gotools_bin_name)
+$$(_gotools_var_name) := $$(_gotools_canonical_bin_path)
+
+ifneq ($(filter ./%,$(2)),)
+# Tool is built from local files. We have to introduce a phony target and let the Go compiler
+# do all the caching.
+.PHONY: $$(_gotools_canonical_bin_path)
+$$(_gotools_canonical_bin_path):
+	@echo "+ $$(notdir $$@)"
+	$$(SILENT)GOBIN="$$(dir $$@)" go install "$(strip $(2))"
+else
+# Tool is specified with version, so we don't take any info from the go.mod file.
+# We install the tool into a location that is version-dependent, and build it via this target. Since the name of
+# the tool under that path is version-dependent, we never have to rebuild it, as it's either the correct version, or
+# does not exist.
+ifneq ($$(_gotools_version),)
+_gotools_versioned_bin_path := $(GOTOOLS_ROOT)/versioned/$$(_gotools_pkg)/$$(_gotools_version)/$$(_gotools_bin_name)
+$$(_gotools_versioned_bin_path):
+	@echo "+ $$(notdir $$@)"
+	$$(SILENT)GOBIN="$$(dir $$@)" go install "$(strip $(2))"
+
+# To make the tool accessible in the canonical location, we create a symlink. This only depends on the versioned path,
+# i.e., only needs to be recreated when the version is bumped.
+$$(_gotools_canonical_bin_path): $$(_gotools_versioned_bin_path)
+	@mkdir -p "$(GOTOOLS_BIN)"
+	$$(SILENT)ln -sf "$$<" "$$@"
+
+else
+
+# Tool is specified with an absolute path without a version. Take info from go.mod file in the respective directory.
+$$(_gotools_canonical_bin_path): $$(_gotools_mod_root)/go.mod
```

Comment:
Why not also `go.sum`?

Reply:
`go.mod` used to be mostly a requirements file. This changed in (I believe) go 1.17, now `go.mod` contains all dependencies, direct and transitive, and `go.sum` is mostly just for checksum verification (`go mod verify`). Therefore, no version of a tool dependency can change without a change in `go.mod`.

I guess you could argue that a tool should be rebuilt when the checksum of the respective module changes. But checksum changes are error cases, they should not normally happen and require human review and intervention (probably resulting in a `go.mod` change) anyway.

Reply:
Thanks, I did not know about that go 1.17 change.

> checksum changes are error cases, they should not normally happen and require human review and intervention

I suppose that's only for packages that are cached in Go Module Mirror. Packages that are consumed directly from git(hub) which are mostly the ones under github.com/stackrox/, can have their tags force-pushed which would change `go.sum` but not change in `go.mod`.

Several probabilities need to be multiplied to get the overall chance of such issue affecting anyone. However, thinking that the issue may happen specifically during local development, it could be very annoying to run into such edge case and see binaries not updated when you expect them to.

No strong feelings as the chance is low and one has an option to just nuke `.gotools/` in anger when things go wrong, which should recover the situation.

Reply:
Yes, the checksum can change in such cases, but I don't know what Go would to do then. I have a hard time imagining that it would silently change the checksum. It would probably ask you to remove the offending checksum entry manually or through a go tool invocation.

However, there is probably no harm at all in specifying `go.sum` in addition (I wanted to avoid a spurious up to date when somebody changed the `go.mod` file w/o running `go mod tidy` afterwards, but nobody says I need to remove `go.mod` as a dep). So added it as well to make it more robust.

As a side note, the `GOPRIVATE` stems from the time when our repos were private. We should probably remove it (will happen in a different PR, though) and get everything through the Go package proxy.

Reply:
Would be nice to remove `GOPRIVATE` because I've seen GitHub throwing 429-s when we try to build the product in CI. Do you plan (have time) doing it yourself?

Reply:
https://github.com/stackrox/stackrox/pull/3402